src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,498 @@
+/*-------------------------------------------------------------------------
+ *
+ * locally_reserved_shared_connection_stats.c
+ *
+ *   Keeps track of the number of reserved connections to remote nodes
+ *   for this backend. The primary goal is to complement the logic
+ *   implemented in shared_connection_stats.c which aims to prevent excessive
+ *   number of connections (typically > max_connections) to any worker node.
+ *   With this locally reserved connection stats, we enforce the same
+ *   constraints considering these locally reserved shared connections.
+ *
+ *   To be more precise, shared connection stats are incremented only with two
+ *   operations: (a) Establishing a connection to a remote node
+ *               (b) Reserving connections, the logic that this
+ *                   file implements.
+ *
+ *   Finally, as the name already implies, once a node has reserved a  shared
+ *   connection, it is guaranteed to have the right to establish a connection
+ *   to the given remote node when needed.
+ *
+ *   For COPY command, we use this fact to reserve connections to the remote nodes
+ *   in the same order as the adaptive executor in order to prevent any resource
+ *   starvations. We need to do this because COPY establishes connections when it
+ *   recieves a tuple that targets a remote node. This is a valuable optimization
+ *   to prevent unnecessary connection establishments, which are pretty expensive.
+ *   Instead, COPY command can reserve connections upfront, and utilize them when
+ *   they are actually needed.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "commands/dbcommands.h"
+#include "distributed/listutils.h"
+#include "distributed/locally_reserved_shared_connection_stats.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_executor.h"
+#include "distributed/placement_connection.h"
+#include "distributed/shared_connection_stats.h"
+#include "distributed/tuplestore.h"
+#include "distributed/worker_manager.h"
+#include "utils/hashutils.h"
+#include "utils/builtins.h"
+
+
+#define RESERVED_CONNECTION_STATS_COLUMNS 4
+
+
+/* session specific hash map*/
+static HTAB *SessionLocalReservedConnectionCounters = NULL;
+
+
+/*
+ * Hash key for connection reservations
+ */
+typedef struct ReservedConnectionCounterHashKey
+{
+	char hostname[MAX_NODE_LENGTH];
+	int32 port;
+	Oid databaseOid;
+} ReservedConnectionCounterHashKey;
+
+/* hash entry for per worker stats */
+typedef struct ReservedConnectionCounterHashEntry
+{
+	ReservedConnectionCounterHashKey key;
+
+	int reservedConnectionCount;
```

Comment:
As discussed:

This could maybe be `bool usedReservation` which starts at false and then becomes true.
 
If it is false, we have a reservation that we can use. 
If it is true, we used the reservation and cannot make more reservations.
If there is no entry in the hash, we can make a reservation.